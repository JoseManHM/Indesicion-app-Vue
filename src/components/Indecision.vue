<template>
    <img v-if="img" :src="img" alt="bg">
    <div class="bg-dark"></div>
    <div class="indecision-container">
        <input type="text" placeholder="Hazme una pregunta" v-model="question">
        <p>Recuerda poner un signo de interrogación (?) al final</p>
        <div>
            <h2>{{question}}</h2>
            <h1 v-show="isValidQuestion">{{respuesta}}</h1>
        </div>
    </div>
</template>

<script>
export default {
    data(){
        return{
            question: '',
            respuesta: '',
            img: '',
            isValidQuestion: false
        }
    },
    watch: {
        question(value, oldValue){
            this.isValidQuestion = false
            console.log({value})
            if(!value.includes('?')) return

            this.isValidQuestion = true
            //TODO: realizar petición HTTP
            this.getAnswer()
        }
    },
    methods:{
        async getAnswer(){
            try{
                this.img = 'https://via.placeholder.com/750'
                this.respuesta = 'Pensando...'
                const { answer, image } = await fetch('https://yesno.wtf/api').then(response => response.json())
                if(answer == 'yes'){
                    this.respuesta = 'Sí!'
                }else if(answer == 'no'){
                    this.respuesta = 'No!'
                }else{
                    this.respuesta = 'Tal vez'
                }
                //this.respuesta = answer
                this.img = image
            }catch(error){
                console.log('getAnswerError: ', error)
                this.respuesta = 'No se pudo cargar el API'
                this.img = null
            }
        }
    }
}
</script>

<style scoped>
    img, .bg-dark{
       height: 100vh;
       left: 0px;
       max-height: 100%;
       max-width: 100%;
       position: fixed;
       top: 0px;
       width: 100vw;
    }
    .bg-dark{
        background-color: rgba(0,0,0,0.4);
    }
    .indecision-container{
        position: relative;
        z-index: 99;
    }
    input{
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus{
        outline: none;
    }
    p{
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }
    h1, h2{
        color: white;
    }
    h2{
        margin-top: 150px;
    }
</style>