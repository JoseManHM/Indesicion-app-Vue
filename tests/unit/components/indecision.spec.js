import { shallowMount, mount } from "@vue/test-utils";
import Indecision from '@/components/Indecision';

describe('Suite de pruebas del componente Indecision', () => {
    let wrapper;
    let consoleLogSpy;

    //Mock para el fetch API
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }))

    beforeEach( () => {
        wrapper = shallowMount(Indecision)

        consoleLogSpy = jest.spyOn(console, 'log')

        jest.clearAllMocks()
    })

    test('Debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Escribir en el input no debe de disparar nada (console.log)', async () => {
        //Spy en el componente
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input =  wrapper.find('input')
        await input.setValue('Hola mundo')

        expect(consoleLogSpy).toHaveBeenCalledTimes(1)
        //expect(getAnswerSpy).toHaveReturnedTimes(0)
        expect(getAnswerSpy).not.toHaveBeenCalled()
        //console.log(wrapper.vm)
    })

    test('Escribir el simbolo ? debe de disparar el getAnswer', async() => {
        //Spy
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('?')

        expect(consoleLogSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalled()
    })

    test('Pruebas en getAnswer', async() => {
        await wrapper.vm.getAnswer()

        const img  = wrapper.find('img')
        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.respuesta).toBe('SÃ­!')
    })

    test('Pruebas en getAnswer - Fallo en el API', async() => {
        
        fetch.mockImplementationOnce(() => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.respuesta).toBe('No se pudo cargar el API')
    })
})