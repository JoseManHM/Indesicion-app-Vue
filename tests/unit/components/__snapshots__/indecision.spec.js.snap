// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suite de pruebas del componente Indecision Debe de hacer match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda poner un signo de interrogaci√≥n (?) al final</p>
  <div>
    <h2></h2>
    <h1 style="display: none;"></h1>
  </div>
</div>
`;
